File: problem_007.go
// -*- compile-command: "go build -o problem_007_go problem_007.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func main() {
        fmt.Println("Project Euler - Problem 7:")
        fmt.Println("What is the 10,001st prime number?\n")

	var sieve [500000]bool
	count := 0
	index := 1

	for count < 10001 {
		index++
		if sieve[index] == false {
			count++
			for i := index; i<500000; i += index {
				sieve[i] = true
			}
		}
	}

        fmt.Println(index)
}

File: problem_007.kt
// -*- compile-command: "kotlinc problem_007.kt -include-runtime -d problem_007.jar" -*-
// Copyright (c) 2017 Michael Caldwell

fun main(args: Array<String>) {
        println("Project Euler - Problem 7:")
        println("What is the 10,001st prime number?\n")

        var count = 1
        val limit = 10001
        var sieve = IntArray(1000000, {_ -> 1})

        var n = 3
        var lastPrime = 2
        while (count < limit) {
                if (sieve[n] == 1) {
                        count++
                        lastPrime = n
                        var i = n*2
                        while (i < 1000000) {
                                sieve[i] = 0
                                i += n
                        }
                }
                n += 2
        }

        println(lastPrime)
}
